shim/src/interpose.c

```diff
@@ -94,50 +95,88 @@ static int (*libc_epoll_create1)(int) = NULL;
+            assert((libc_bind = dlsym(RTLD_NEXT, "bind")) != NULL);
+            assert((libc_connect = dlsym(RTLD_NEXT, "connect")) != NULL);
+            assert((libc_fcntl = dlsym(RTLD_NEXT, "fcntl")) != NULL);
+            assert((libc_listen = dlsym(RTLD_NEXT, "listen")) != NULL);
+            assert((libc_accept4 = dlsym(RTLD_NEXT, "accept4")) != NULL);
+            assert((libc_accept = dlsym(RTLD_NEXT, "accept")) != NULL);
+            assert((libc_getsockopt = dlsym(RTLD_NEXT, "getsockopt")) != NULL);
+            assert((libc_setsockopt = dlsym(RTLD_NEXT, "setsockopt")) != NULL);
+            assert((libc_getsockname = dlsym(RTLD_NEXT, "getsockname")) != NULL);
+            assert((libc_getpeername = dlsym(RTLD_NEXT, "getpeername")) != NULL);
+            assert((libc_read = dlsym(RTLD_NEXT, "read")) != NULL);
+            assert((libc_recv = dlsym(RTLD_NEXT, "recv")) != NULL);
+            assert((libc_recvfrom = dlsym(RTLD_NEXT, "recvfrom")) != NULL);
+            assert((libc_recvmsg = dlsym(RTLD_NEXT, "recvmsg")) != NULL);
+            assert((libc_readv = dlsym(RTLD_NEXT, "readv")) != NULL);
+            assert((libc_pread = dlsym(RTLD_NEXT, "pread")) != NULL);
+            assert((libc_write = dlsym(RTLD_NEXT, "write")) != NULL);
+            assert((libc_send = dlsym(RTLD_NEXT, "send")) != NULL);
+            assert((libc_sendto = dlsym(RTLD_NEXT, "sendto")) != NULL);
+            assert((libc_sendmsg = dlsym(RTLD_NEXT, "sendmsg")) != NULL);
+            assert((libc_writev = dlsym(RTLD_NEXT, "writev")) != NULL);
+            assert((libc_pwrite = dlsym(RTLD_NEXT, "pwrite")) != NULL);
+            assert((libc_close = dlsym(RTLD_NEXT, "close")) != NULL);
+            assert((libc_epoll_create = dlsym(RTLD_NEXT, "epoll_create")) != NULL);
+            assert((libc_epoll_create1 = dlsym(RTLD_NEXT, "epoll_create1")) != NULL);
+            assert((libc_epoll_ctl = dlsym(RTLD_NEXT, "epoll_ctl")) != NULL);
+            assert((libc_epoll_wait = dlsym(RTLD_NEXT, "epoll_wait")) != NULL);
+
+            __sync_fetch_and_sub(&in_init_libc, 1);
+            MEM_BARRIER();
+            initialized_libc = 1;
+        }
+        else
+        {
+            while (initialized_libc == 0)
+            {
+                sched_yield();
+            }
+            MEM_BARRIER();
+        }
+    }
 }
 
 static void init(void)
 {
-    if (!initialized)
+    int ret;
+
+    if (initialized == 0)
     {
-        if (__init() != 0)
-            abort();
+        if (__sync_fetch_and_add(&in_init, 1) == 0)
+        {
+            ret = __init();
+            if (ret != 0 && ret != EEXIST)
+                abort();
 
-        initialized = true;
+            __sync_fetch_and_sub(&in_init, 1);
+            MEM_BARRIER();
+            initialized = 1;
+        }
+        else
+        {
+            while (initialized == 0)
+            {
+                sched_yield();
+            }
+            MEM_BARRIER();
```

Comment:
Shouldn't the `MEM_BARRIER` be in the loop?

Reply:
The variable `initialized` is declared as volatile and from my understanding this should prevent the compiler from optimizing things related to this variable and makes sure to fetch the updated value instead of an outdated cached version of it. The `MEM_BARRIER` at the end is to prevent instructions after `init()` from being reordered to somewhere during the initialization code inside `init()`.

Reply:
Ok.